### PR TITLE
[PLT-7574] Update and Simplify Compatibility Check and Replace UA Dependency

### PR DESCRIPTION
**Pull request by moonmeister**
#### Summary
Changing to only check Safari and Edge, Updating UA check dependency to better library.

See: https://pre-release.mattermost.com/core/pl/xou45eqebtro3qx4euzm3nho5w

#### Ticket Link
Current:
https://mattermost.atlassian.net/browse/PLT-7574

Browser Compatibility Check Implementation:
https://github.com/mattermost/mattermost-server/pull/6945
https://mattermost.atlassian.net/browse/PLT-959

### CODE REVIEW

**Comment by moonmeister** [comment]
I wasn't able to test the Mobile browser check, if we get a test server setup I should be able to do that.

**Comment by hmhealey** [comment]
Since the build server is still having issues with the migration, I set up my test server to point to this branch (assuming it doesn't die as well since it's been having issues)

http://harrison.spinmint.com:8065/

**Comment by hmhealey** [comment]
Here's a sign up link for the server http://harrison.spinmint.com:8065/signup_user_complete/?id=g6q63r8jepf5jx6e9s9okcwsdw

**Comment by moonmeister** [comment]
Okay, so the desktop app  (I'm on a mac) detected as an app and skipped the Browser check. My mobile browser detected as mobile and skipped browser check. I tried the mobile apps (classic and beta) and I think they are detecting as "ELB-HealthChecker 2.0". Not sure why but I didn't have any issues with an incompatible dialog, so I think everything is working correctly now. I rebased the branch and it's rebuilding now.

**Comment by hmhealey** [comment]
ELB-HealthChecker 2.0 is an AWS thing that's just checking that the test server is online. The Classic apps would have a different user agent, but that doesn't really matter if they're working after your changes. The Mattermost/Mattermost Beta apps don't need to be checked since they aren't browser-based and therefore won't be hitting your code at all

**Comment by moonmeister** [comment]
That's what I was thinking but I couldn't get the classic app to trigger a browser version debug log that read anything other than that aws thing. Buy the app worked so I suppose that's the most important thing.

**Comment by moonmeister** [comment]
Ah. I know the problem. I have three if/else-if statements. The classic app doesn't result true on any of them and thus never reports the client. I should fix that.

File: web/web.go [hunk 1]
```diff
@@ -96,7 +96,14 @@ func CheckBrowserCompatability(ua *user_agent.UserAgent) bool {
...
 func root(c *api.Context, w http.ResponseWriter, r *http.Request) {
-	if !CheckBrowserCompatability(user_agent.New(r.UserAgent())) {
+	agentString := r.UserAgent()
+	ua := user_agent.New(agentString)
+
+	if strings.Contains(agentString, "Mattermost") {
+		l4g.Debug("Detected Browser: Mattermost App")
+	} else if ua.Mobile() {
+		l4g.Debug("Detected Browser: Mobile Browser")
+	} else if !CheckBrowserCompatability(ua) {
```
Comment (hmhealey) [inline web/web.go]: While you're changing that to always print the browser, do you mind changing CheckBrowserCompatibility to take the user agent string and check for the app inside the function, so that it's all in one place? Having some cases in this if statement return and others not makes it a bit hard to follow the flow. You can also have CheckBrowserCompatibility print the name of the browser in each case

**Comment by hmhealey** [comment]
Also, let me know when you make any changes, and I can update the test server

**Comment by moonmeister** [comment]
Read here for my current issue: https://github.com/mssola/user_agent/issues/39

Basically I can add in a bunch of exceptions for Franz and the Mattermost app, or just electron as a whole, but I don't know how many other third party apps are out there detecting strangely. 

The way browser compatibility check only results as incompatible if it recognizes the browser name and the version is unsupported. So, unknown browsers, Vivaldi for example, are allowed and only browsers included in the `browserMinimumSupported` map are version checked. Where this fails (in the case of Electron) is that it's detecting as Safari, but the user_agent tool is not reporting the correct version. Basically the tool tries to do it's best but fails when it hits a UA string it hasn't been programmed to understand. It fails poorly as opposed to telling you it doesn't recognize the UA string.

IMO:
1. We make an exception for Mobile if mobile browser version truly doesn't matter.
2. We get the user_agent tool to detect Electron correctly and set a minimum required version for that (or just detect Electron via string if all electron versions work).
3. We modify the error page to include where people can ask for client compatibility support. Using UA string as we currently are is going to result in incompatibility reports it seems when the user_agent tool fails. If we can make it simple for users to request their client be checked for compatibility then it's not as big of a deal. I guess some of this goes back to what Mattermost wants to support too. 

Other thing: I'll look into making some contributions to the user_agent tool but we might think about forking it to make it less error prone and fit our needs.

**Comment by moonmeister** [comment]
@hmhealey @jasonblais I'm taking a break on this for now. I've run into so many issues with the [user_agent checker](https://github.com/mssola/user_agent). It's super delicate. I don't think it's production ready. Beyond that everything I've read online says don't do Browser Version checking unless you absolutely have to, MDN has several recommendations on alternatives. 

Basically I'm starting to think there is fundamental issue with the solution (UA string checking) for our problem (browser compatibility for the sake of features? and security?). I think there needs to be some discussion around defining the actual issue we're trying to solve, then once we've clarified that we can re-attempt to create a solution. Even if we think this is the right solution I think there's some major issues with it that people need to be ready to accept. 

I'll cue a discussion  item for the dev meeting on Sept. 20th. I'm happy to jump on a call with the relevant people before then or instead of discussing the issue there.

FInally, Until this discussion happens and we can solve this problem correctly I might suggest reverting the merge #6945. It is probably affecting people's dev work and definitely is not ready for v4.3.

**Comment by esethna** [comment]
Thanks @moonmeister, agree we should take a step back and discuss. Originally the issue was to simply change the text appearing on the unsupported browser page to link to the supported browsers: 
image


I understand from the discussions in these PRs that our browser detection could use some improvement. I will defer to @hmhealey on the technical side, but anything beyond the text change should probably be considered in a different ticket and discussed in the dev meeting, given that we've hit issues with the implementation as we've attempted it. 

Appreciate your effort here @moonmeister,

**Comment by moonmeister** [comment]
@esethna @hmhealey I don't know how hard it would be but we could revert the detection changes and leave the UI changes. Especially now that the repo has been split since this change was merged.

**Comment by hmhealey** [comment]
I don't think it should be too difficult to only revert part of them. Luckily, I think your changes only affected the serverside code

**Comment by moonmeister** [comment]
I'm pretty sure there were some web app changes. But do as you will. Once we talk on Wednesday we'll have a better idea of how much will be changing from the current system.

**Comment by moonmeister** [comment]
@esethna The discussion basically decided we need to check to see why/how this "Incompatible browser" list was generated. What's the actual problems with these older browser versions? Corey seemed to think that the only reason this was ever implemented was to block IE < 11. Once we better understand why these browsers are considered incompatible then we decide the best way to move forward.

**Comment by jasonblais** [comment]
For Edge it was this: https://mattermost.atlassian.net/browse/PLT-7091

Basically, Edge has some issues with at least the following:
- Error messages on login (bad password, etc.)
- MFA enforcement (not 100% sure but likely)
- Account creation when email verification is forced
- Possibly some error messages for other login related activities

It was caused by https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/8546263/

**Comment by hmhealey** [comment]
Ah, so the fetch API had problems in that release. I think we have a polyfill so it would work on older versions, but not on that specific version

**Comment by hmhealey** [comment]
I'm looking for other browsers and we dropped:
- Safari 7 because it was old (https://mattermost.atlassian.net/browse/PLT-941)
- Safari 8 because it had some issues (https://mattermost.atlassian.net/browse/PLT-1136, https://github.com/mattermost/mattermost-server/pull/1418)
- Edge < 40 because of the above issue (https://mattermost.atlassian.net/browse/PLT-7091)
- Firefox < 52 because of this issue (https://github.com/mattermost/mattermost-server/issues/6428, https://mattermost.atlassian.net/browse/PLT-6650)
- IE11 compatibility view for basically the same reason we don't support IE10 and lower

In my mind, I don't think we need to prevent users from opening the page on those browsers. The app may work in most cases, but if it doesn't, we can just point to the docs to say that we don't support them. It's not a perfect user experience, but most of those browsers are relatively old (old versions of IE, Safari) or are modern browsers that force updates every other week (Edge, Firefox, Chrome). I'd propose we keep the updated text on that page, but we revert the changes to browser detection to only show it in those rare cases since this is a lot of effort for less than 1% of our audience

**Comment by moonmeister** [comment]
@hmhealey Sounds good. Can we define a specific list of browsers we want to send this error on? 

IE < 11
Safari?

**Comment by moonmeister** [comment]
@hmhealey Also We should still update our dependency! It's bad how bad it is. I took a look at both:

https://github.com/xojoc/useragent
https://github.com/avct/uasurfer

Both seemed better uasurfer had really clear code but detected Edge as IE. WE could submit a PR or Fork it. xojoc/useragent's code was less clear but did a good job of detection.

**Comment by moonmeister** [comment]
Once #7497 is merged I'll rebase this and make sure everything is good to go. This won't build though untill the new dependency is added to the pr. Is that something @crspeller needs to do?

**Comment by moonmeister** [comment]
Okay, it's been rebased. I just not sure the protocol for adding the new dependency to this PR, I'll wait to hear.

**Comment by jasonblais** [comment]
@hmhealey @esethna Where are we on this?

**Comment by hmhealey** [comment]
I think we'll want to remove the debug messages, but it looks good to me barring whatever is causing the build failure. I think it might just need to be rebased, but I'll close and reopen it to trigger a build

**Comment by hmhealey** [comment]
@moonmeister Looks like the new useragent library needs to be added to the vendor directory so that the PR will build
```
env GOOS=linux GOARCH=amd64 go install  -ldflags "-X github.com/mattermost/mattermost-server/model.BuildNumber=690 -X 'github.com/mattermost/mattermost-server/model.BuildDate=Wed Nov 15 16:42:46 UTC 2017' -X github.com/mattermost/mattermost-server/model.BuildHash=709066f81b7acc3926d01f939a17140728e1b3a9 -X github.com/mattermost/mattermost-server/model.BuildHashEnterprise=cef4ad8a312cbce91e5cab11cbc5b4ad5f254d21 -X github.com/mattermost/mattermost-server/model.BuildEnterpriseReady=true" ./cmd/platform
web/web.go:10:2: cannot find package "xojoc.pw/useragent" in any of:
	/home/ubuntu/workspace/mattermost-platform-pr-split/src/github.com/mattermost/mattermost-server/vendor/xojoc.pw/useragent (vendor tree)
	/usr/local/go/src/xojoc.pw/useragent (from $GOROOT)
	/home/ubuntu/workspace/mattermost-platform-pr-split/src/xojoc.pw/useragent (from $GOPATH)
make: *** [build-linux] Error 1
```

**Comment by moonmeister** [comment]
@hmhealey Agreed. Last I heard the community wasn't suppose to modify dependencies and someone on the MM team would add the dependency to the PR. How shall I proceed?

File: web/web.go [hunk 2]
```diff
@@ -7,13 +7,14 @@ import (
 	"net/http"
 	"strings"
 
+	"xojoc.pw/useragent"
```
Comment (jasonblais) [inline web/web.go]: @moonmeister Sorry, this repo uses a GPLv3 license which is not compatible with the Mattermost software.

Comment (jasonblais) [reply web/web.go]: Any alternative library we can consider?

Comment (moonmeister) [reply web/web.go]: Oh, Ok. What is compatible?

Comment (jasonblais) [reply web/web.go]: @moonmeister Here's a list that's a good start. https://www.apache.org/legal/resolved.html#category-a

**Comment by moonmeister** [comment]
@hmhealey @jasonblais I'll look into other options but short of rolling our own we might be stuck with what we got. Fortunately, it works for what we are currently doing.

File: web/web.go (hunk 2 above)
Comment (moonmeister) [reply web/web.go]: According to the Licence it seems only the webapp is under Apache 2.0. the Server seems to be under GNU AGPL v.3.0...Maybe I'm missing something.

**Comment by moonmeister** [comment]
Okay, @hmhealey @jasonblais! I've changed the dependency to https://github.com/avct/uasurfer (Apache 2.0). Tests are passing and debug messages removed. I still haven't modified the vendor directory as I believe @crspeller has to do so. I'm happy to do it if you want me to do it, just let me know. Thanks.

**Comment by jasonblais** [comment]
Great, thank you! @esethna Can you help coordinate and make sure @crspeller includes this dependancy in the next dependancy update (likely next week).

**Comment by moonmeister** [comment]
@crspeller @esethna @jasonblais Where we at on this?

**Comment by grundleborg** [comment]
*Closing and reopening to trigger Jenkins**

**Comment by moonmeister** [comment]
@crspeller @esethna @jasonblais Where we at on this?

**Comment by esethna** [comment]
@moonmeister looks like the build is failing, can you help investigate?

**Comment by moonmeister** [comment]
@esethna has anyone added the dependency?

**Comment by esethna** [comment]
@moonmeister, just checked with @hmhealey, he's helping add the dependency

File: web/web.go [hunk 3]
```diff
@@ -65,26 +65,27 @@ func pluginHandler(config model.ConfigFunc, handler http.Handler) http.Handler {
 	})
 }
 
-var browsersNotSupported string = "MSIE/8;MSIE/9;MSIE/10;Internet Explorer/8;Internet Explorer/9;Internet Explorer/10;Safari/7;Safari/8"
+// Due to the complexities of UA detection and the ramifications of a misdetection only older Safari and IE browsers throw incopatability errors.
```
Comment (ccbrown) [inline web/web.go]: "incopatability"

File: web/web_test.go [hunk 4]
```diff
@@ -153,4 +153,104 @@ func TestMain(m *testing.M) {
...
+		t.Error("Fail: Safari 8 incorrectly passed browser compatibility!")
+	}
+
+	//Safari Mobile
+	ua = "Mozilla/5.0 (iPhone; CPU iPhone OS 9_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13B137 Safari/601.1"
+	if result := CheckClientCompatability(ua); result == true {
+		t.Log("Pass: Safari Mobile passed browser compatibility.")
+	} else {
+		t.Error("Fail: Safari Mobile User Agent Test failed!")
+	}
```
Comment (ccbrown) [inline web/web_test.go]: Some table-driven-testing would make this a lot nicer:

```go
func TestCheckClientCompatability(t *testing.T) {
	for name, tc := range map[string]struct {
		UserAgent string
		Expected  bool
	}{
		"Firefox 40.1": {
			UserAgent: "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.1",
			Expected:  true,
		},
		"Chrome 60": {
			UserAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36",
			Expected:  true,
		},
		"Chrome Mobile": {
			UserAgent: "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Mobile Safari/537.36",
			Expected:  true,
		},
		"Classic app": {
			UserAgent: "Mozilla/5.0 (Linux; Android 8.0.0; Nexus 5X Build/OPR6.170623.013; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/61.0.3163.81 Mobile Safari/537.36 Web-Atoms-Mobile-WebView",
			Expected:  true,
		},
		"Mattermost App 3.7.1": {
			UserAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Mattermost/3.7.1 Chrome/56.0.2924.87 Electron/1.6.11 Safari/537.36",
			Expected:  true,
		},
		"Franz 4.0.4": {
			UserAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Franz/4.0.4 Chrome/52.0.2743.82 Electron/1.3.1 Safari/537.36",
			Expected:  true,
		},
		"Edge 14": {
			UserAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393",
			Expected:  true,
		},
		"IE 11": {
			UserAgent: "Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko",
			Expected:  true,
		},
		"IE 9": {
			UserAgent: "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 7.1; Trident/5.0)",
			Expected:  false,
		},
		"Safari 9": {
			UserAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Safari/604.1.38",
			Expected:  true,
		},
		"Safari 8": {
			UserAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/600.7.12 (KHTML, like Gecko) Version/8.0.7 Safari/600.7.12",
			Expected:  false,
		},
		"Safari Mobile": {
			UserAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 9_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13B137 Safari/601.1",
			Expected:  true,
		},
	} {
		t.Run(name, func(t *testing.T) {
			assert.Equal(t, tc.Expected, CheckClientCompatability(tc.UserAgent))
		})
	}
}
```

Comment (moonmeister) [reply web/web_test.go]: Done. Good Idea!

File: web/web.go (hunk 3 above)
Comment (moonmeister) [reply web/web.go]: Oops.

**Comment by hmhealey** [comment]
@moonmeister If you pull the latest master, I've added the uasurfer dependency

**Comment by moonmeister** [comment]
@jasonblais @esethna Anything else left on this?

**Comment by esethna** [comment]
@moonmeister can you list out the browsers/versions where the unsupported browser page is expected to appear?

Passing this along to @lindalumitchell to help test now that the dependency is updated.

**Comment by moonmeister** [comment]
Anything less than the browser version called out in `browserMinimumSupported`. Browsers not listed will always pass.

Currently:
```
var browserMinimumSupported = map[string]int{
     "BrowserIE":     11,
     "BrowserSafari": 9,
 }
```

So, IE >= 11 and Safari >= 9 will succeed but anything < will fail and be sent to the compatibility page. That functionality hasn't changed, it's just be tested with a package the accurately detects browser+version, the code has been simplified, and tests added.

**Comment by esethna** [comment]
@moonmeister sorry for the delay on this PR, we are just figuring out how to test this thoroughly.

**Comment by moonmeister** [comment]
@esethna Given I've written to have tests could you download some massive spreadsheet of user agents and run it through the tests to see if they all pass except for old IE and Safari? I'm sure you could find a comprehensive list.

**Comment by hmhealey** [comment]
I've gone through and double checked on a bunch of different browsers as well as the desktop app and classic mobile app, and it looks good on all of them. Sorry again about the delay in getting to that, @moonmeister.

@ccbrown Can you give this a quick second review?

File: web/web_test.go [hunk 5]
```diff
@@ -156,4 +156,35 @@ func TestMain(m *testing.M) {
...
+		{"Safari 9", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Safari/604.1.38", true},
+		{"Safari 8", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/600.7.12 (KHTML, like Gecko) Version/8.0.7 Safari/600.7.12", false},
+		{"Safari Mobile", "Mozilla/5.0 (iPhone; CPU iPhone OS 9_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13B137 Safari/601.1", true},
+	}
+	for _, browser := range uaTestParameters {
+		if result := CheckClientCompatability(browser.UserAgent); result == browser.Result {
+			t.Logf("Pass: %s passed browser test.", browser.Name)
+		} else {
+			t.Errorf("Fail: %s User Agent Test failed!", browser.Name)
+		}
```
Comment (ccbrown) [inline web/web_test.go]: This would be a lot better with subtests ([t.Run](https://golang.org/pkg/testing/#T.Run)).

Comment (hmhealey) [reply web/web_test.go]: I'm going to go remove the old dependency, so I might grab that while I'm in there

**Comment by hmhealey** [comment]
Thanks again for the PR, @moonmeister!

**Comment by moonmeister** [comment]
Woohoo!!! Thanks Guys! :smile: :fireworks: :champagne:
